aten/src/ATen/native/EmbeddingBag.cpp

```diff
@@ -32,11 +32,16 @@ namespace native {
 template<typename scalar_t>
 scalar_t dot_impl(int64_t n, scalar_t *x, int64_t incx, scalar_t *y, int64_t incy);
 
-static void make_offset2bag(const Tensor &offsets, const Tensor &indices, Tensor& offset2bag) {
-  offset2bag.index_add_(
-      0, offsets, at::ones_like(offsets, LEGACY_CONTIGUOUS_MEMORY_FORMAT)); // offset2bag = [1 0 1 0 1]
-  offset2bag[0] -= 1;                     // offset2bag = [0 0 1 0 1]
-  offset2bag = offset2bag.cumsum(0);     // offset2bag = [0 0 1 1 2]
+static void make_offset2bag(const Tensor &offsets, Tensor& offset2bag) {
+  AT_DISPATCH_INDEX_TYPES(offsets.scalar_type(), "make_offset2bag", [&] () {
```

Comment:
@qizzzh The reason that XLA test failed is here data storage is directly accessed. From native_functions.yaml _embedding_bag_backward (which calls make_offset2bag) should be device generic (cannot assume storage exists). There're two ways to fix:
1. Use at:: functions instead of directly accessing storage in make_offset2bag
2. Move the old _embedding_bag_backward implementation to Math key and add dispatch section for it in native_functions.yaml. The key here is after this change, your kernel no longer works for all backends. More details can found https://github.com/pytorch/pytorch/tree/master/aten/src/ATen/native#choosing-the-right-dispatch-keyword
Either solution works for me, the first one might be a little less work I think.

Reply:
Thanks for the explanation. I guess I'll just make index_add support int32 indices so we keep the original code. index_add is pretty self-contained and I hope it would be fine.